crates/miden-objects/src/account/storage/map/hashed_storage_map_key.rs

```diff
@@ -0,0 +1,37 @@
+use core::fmt;
+
+use miden_crypto::Word;
+
+use crate::Felt;
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy, PartialOrd, Ord)]
+pub struct HashedStorageMapKey(Word);
+
+impl HashedStorageMapKey {
+    /// Creates a new [`HashedStorageMapKey`] from the given [`Word`] **without performing
+    /// validation**.
+    ///
+    /// ## Warning
+    ///
+    /// This function **does not check** whether the provided `Word` represents a valid
+    /// fungible or non-fungible asset key.
+    pub fn new_unchecked(value: Word) -> Self {
+        Self(value)
+    }
+
+    pub fn inner(&self) -> Word {
+        self.0
+    }
+
+    /// Returns the leaf index of a map key.
+    pub fn hashed_map_key_to_leaf_index(&self) -> Felt {
```

Comment:
This could look just like `AssetVault::to_leaf_index`:

https://github.com/0xMiden/miden-base/blob/0cd80afdbd9b9d26e27341fad895063ac53eacdc/crates/miden-objects/src/asset/vault/vault_key.rs#L64-L67